core/src/main/scala/org/apache/spark/sql/delta/stats/StatisticsCollection.scala

```diff
@@ -389,7 +384,306 @@ trait StatisticsCollection extends DeltaLogging {
+    val cols = statsColPaths.groupBy(_.head).mapValues(_.map(_.tail))
+    val newSchema = schema.flatMap { field =>
+      cols.get(field.name).flatMap { paths =>
+        field.dataType match {
+          case _ if paths.forall(_.isEmpty) =>
+            val fullPath = (parentPath:+ field.name).mkString(".")
+            def convertToPhysicalName(field: StructField): StructField = {
+              // If mapping mode is NoMapping or the dataSchemaName already contains the mapped
+              // column name, the schema mapping can be skipped.
+              if (mappingMode == NoMapping || schemaNames.contains(fullPath)) return field
+              val physicalName = field.metadata.getString("delta.columnMapping.physicalName")
+              field.dataType match {
+                case structType: StructType =>
+                  val newDataType = StructType(structType.map(convertToPhysicalName))
+                  field.copy(name = physicalName, dataType = newDataType)
+                case _ => field.copy(name = physicalName)
+              }
+            }
+            Some(convertToPhysicalName(field))
+          case fieldSchema: StructType =>
+            val newPath = parentPath:+ field.name
+            // Recurse into the child
+            filterSchema(schemaNames, fieldSchema, paths.filterNot(_.isEmpty), mappingMode, newPath)
+              .map(newFieldSchema => field.copy(dataType = newFieldSchema))
+          case _ =>
+            // Filter expected a nested field and this isn't nested. No match
+            None
+        }
+      }
+    }
+    if (newSchema.nonEmpty) Some(StructType(newSchema.toArray)) else None
+  }
+
+  /**
+   * Computes the set of columns to be used for stats collection on a table. Specific named columns
+   * take precedence, if provided; otherwise the first numIndexedColsOpt are extracted from the
+   * schema.
+   */
+  def getIndexedColumns(
+      schemaNames: Seq[String],
+      spec: DeltaStatsColumnSpec,
+      schema: StructType,
+      mappingMode: DeltaColumnMappingMode): StructType = {
+    spec.deltaStatsColumnNamesOpt
+      .map { indexedColNames =>
+        val indexedColPaths = indexedColNames.map(_.nameParts)
+        filterSchema(schemaNames, schema, indexedColPaths, mappingMode).getOrElse(new StructType())
+      }
+      .getOrElse {
+        val numIndexedCols = spec.numIndexedColsOpt.get
+        if (numIndexedCols < 0) {
+          schema // negative means don't truncate the schema
+        } else {
+          truncateSchema(schema, numIndexedCols)._1
+        }
+      }
+  }
+
+  /**
+   * Generates a truncated data schema for stats collection.
+   * @param schema the original data schema
+   * @param indexedCols the maximum number of leaf columns to collect stats on
+   * @return truncated schema and the number of leaf columns in this schema
+   */
+  def truncateSchema(schema: StructType, indexedCols: Int): (StructType, Int) = {
```

Comment:
this was private before.. does it need to be public?

Reply:
Made it private.